fiftyone/core/stages.py

```diff
@@ -1795,9 +1889,66 @@ def _get_frames_list_field_only_matches_expr(field):
     return F("frames").reduce(VALUE + F(field).length()) > 0
 
 
-def _get_list_field_mongo_filter(filter_arg, prefix="$this"):
+def _get_trajectories_filter(sample_collection, field, filter_arg):
+    label_type = sample_collection._get_label_field_type(field)
+    path, is_frame_field = sample_collection._handle_frame_field(field)
+
+    if not is_frame_field:
+        raise ValueError(
+            "Filtering trajectories is only supported for frame fields"
+        )
+
+    if issubclass(label_type, (fol.Detections, fol.Polylines, fol.Keypoints)):
+        path += "." + label_type._LABEL_LIST_FIELD
+        cond = _get_list_trajectory_mongo_filter(filter_arg)
+        filter_expr = (F("index") != None) & foe.ViewExpression(cond)
+        reduce_expr = VALUE.extend(
+            (F(path) != None).if_else(
+                F(path).filter(filter_expr).map(F("index")), [],
+            )
+        )
+    elif issubclass(label_type, (fol.Detection, fol.Polyline, fol.Keypoint)):
+        cond = _get_trajectory_mongo_filter(filter_arg)
+        filter_expr = (F("index") != None) & foe.ViewExpression(cond)
+        reduce_expr = (
+            F(path)
+            .apply(filter_expr)
+            .if_else(VALUE.append(F(path + ".index")), VALUE)
+        )
+    else:
+        raise ValueError(
+            "Cannot filter trajectories for field '%s' of type %s"
+            % (field, label_type)
+        )
+
+    indexes_expr = F("frames").reduce(reduce_expr, [])
+
+    set_pipeline = [{"$set": {"_indexes": indexes_expr.to_mongo()}}]
```

Comment:
Converting the index lists to a set with `{ $setIntersection: [ "$_indexes", "$_indexes" ] }` or similar might be a small optimization. Or only creating index sets with `$addToSet`.

Reply:
Oh good point.

I tried using set arithmetic both during the `$reduce` and at the end like you mentioned, and they were actually pretty similar in runtime, but, since `$addToSet` isn't directly supported in `$reduce`, the `$setUnion` workflow likely wouldn't scale as well for very large datasets, so I went with the "at the end" option.

This change makes the pipeline ~35% faster on my test dataset. Thanks for the idea! 💪